### PR TITLE
Add SchemaVisitor and QueryVisitor traits and functions

### PR DESCRIPTION
**Pull request by davidpdrsn**
Fixes https://github.com/graphql-rust/graphql-parser/issues/25

Implementation is based upon the description here https://github.com/rust-unofficial/patterns/blob/master/patterns/visitor.md

### CODE REVIEW

**Comment by tomhoule** [comment]
It looks good to me on the whole! It's unsollicited, but if nobody else has the time, I could do a more in-depth review.

**Comment by davidpdrsn** [comment]
@tomhoule that would be great 😊

File: src/query/query_visitor.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        Fragment(inner) => {
+            visitor.visit_fragment_definition(inner);
+            walk_fragment_definition(visitor, inner);
+        },
+    }
+}
+
+fn walk_fragment_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast FragmentDefinition) {
+    visitor.visit_fragment_definition(node);
+}
```
Comment (tomhoule) [inline src/query/query_visitor.rs]: Would it make sense to further walk down the fragment fields?

File: src/query/query_visitor.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    visitor.visit_field(node)
+}
+
+fn walk_fragment_spread<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast FragmentSpread) {
+    visitor.visit_fragment_spread(node)
+}
+
+fn walk_inline_fragment<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast InlineFragment) {
+    visitor.visit_inline_fragment(node);
+}
```
Comment (tomhoule) [inline src/query/query_visitor.rs]: Same as for fragments, would it make sense to visit the selection set?

Comment (davidpdrsn) [reply src/query/query_visitor.rs]: Addressed in cb71a7e 👍

File: src/query/query_visitor.rs (hunk 1 above)
Comment (davidpdrsn) [reply src/query/query_visitor.rs]: Addressed in db9310e 👍

**Comment by LegNeato** [comment]
@tailhook will leave it to you to look at and land.

File: src/query/query_visitor.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+fn walk_selection<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Selection) {
+    use super::ast::Selection::*;
+
+    visitor.visit_selection(node);
+    match node {
+        Field(inner) => {
+            visitor.visit_field(inner);
+            walk_field(visitor, inner);
```
Comment (tailhook) [inline src/query/query_visitor.rs:203]: I don't understand this pattern. You first do `visit_field` and then `walk_field` which calls `visit_field` again. Same with fragments. Am I misunderstanding something?

File: src/query/query_visitor.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+//!                 country {
+//!                     id
+//!                 }
+//!             }
+//!         }
+//!     "#).expect("Failed to parse query");
+//!
+//!     walk_document(&mut number_of_type, &doc);
+//!
+//!     assert_eq!(number_of_type.count, 2);
```
Comment (tailhook) [inline src/query/query_visitor.rs:49]: This doesn't look right. There are four fields if visiting recursively: `query, id, country, id` and single one if not (`users`). Currently it looks like you just counting `users` twice.

**Comment by tailhook** [comment]
> It looks like there are multiple mistakes with visiting recursively.

Or was non-recursive visitor intentional?

File: src/query/query_visitor.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+//! }
+//!
+//! impl<'ast> QueryVisitor<'ast> for FieldsCounter {
+//!     fn visit_field(&mut self, node: &'ast Field) {
+//!         self.count += 1
+//!     }
+//! }
+//!
+//! fn main() {
+//!     let mut number_of_type = FieldsCounter::new();
```
Comment (spawnia) [inline src/query/query_visitor.rs:34]:
```suggestion
//!     let mut fields_counter = FieldsCounter::new();
```

File: src/query/query_visitor.rs (hunk 3 above)
Comment (vladinator1000) [reply src/query/query_visitor.rs:203]: I have a similar question. It seems that every walk function calls visit_node and then then calls a walk_node function that calls the same visit_node. For example `walk_subscription` calls `visitor.visit_selection_set` and then cals `walk_selection_set` which in turn calls `visitor.visit_selection_set` a second time. 

Is that how it's supposed to work? I thought you only need to visit once.

**Comment by vladinator1000** [comment]
Hi y'all, I'm gonna try another experimental flavor of an implementation for this, following the [`graphql-js` interface](https://github.com/graphql/graphql-js/blob/8ca3d896794a866e3de6fcba3109874cd84674d8/src/language/visitor.ts#L20). Here is my reasoning for this:

1. `enter` would let you manipulate a node of the AST tree before entering its subtree
2. `leave` would let you manipulate a node of the AST tree after its subtree has been visited
3. It would pave the way for schema-aware visitors using a [`TypeInfo`](https://github.com/graphql/graphql-js/blob/8ca3d896794a866e3de6fcba3109874cd84674d8/src/utilities/TypeInfo.ts#L46) struct. Here's some [example usage](https://github.com/graphql/graphql-js/blob/8ca3d896794a866e3de6fcba3109874cd84674d8/src/utilities/__tests__/TypeInfo-test.ts#L151).

The biggest use-case of a system like this in my case is doing custom validation of a query based on a schema, this API lets you do that.

**Comment by dotansimha** [comment]
We implemented a visitor (with type info) for both schema and queries in https://github.com/dotansimha/graphql-tools-rs , also a (almost) spec-compliant validation phase. 

Thanks for this PR, it was a great inspiration!

**Comment by davidpdrsn** [comment]
@dotansimha why did you post this in 3 different places? 😅

**Comment by dotansimha** [comment]
> @dotansimha why did you post this in 3 different places? 😅

I saw these PRs are duplicated anyway, so I hope it would help someone :) I'll delete if that's a problem ;) didn't want to annoy with notification :x

**Comment by davidpdrsn** [comment]
I'll close this for now. Its been a long time and I no longer have bandwidth to follow this through.

**Pull request closed without merging**